### PR TITLE
Add X-Forwarded-Client-Certificate (XFCC) support

### PR DESCRIPTION
**Pull request by jwm**
## Description
Adds a configurable SNI hostname on originated TLS connections.

## Related Issues
#1543

## Testing
We're running this in production, but with Ambassador 0.75. I ported it to the current `master`, but hit a wall running the test suite (see #2052). IIRC, the test was working at one point, but broke when I rebased onto an Ambassador release with Python type checking enabled. It looked like there was a legit type problem at the time, but I haven't touched the test recently.

## Todos
- [ ] Tests
- [X] Documentation

### CODE REVIEW

**Comment by jwm** [comment]
notstale

**Comment by kflynn** [comment]
Hey @jwm -- want to take a crack at conflict resolution here? I can take a look at typing after that... thanks!!

**Comment by jwm** [comment]
@kflynn sure, done!

Also, to specifically flag it: the Python test here is a copy pasta of `ServerNameTest` and doesn't meaningfully test this change. I haven't looked at the test subsystem recently -- if you've finished that refactor to make it easier to test (or could give some quick info/pointers to docs on how to go about it), I can poke at that hopefully this week.

File: python/schemas/v1/TLSContext.schema [hunk 1]
```diff
@@ -34,7 +34,10 @@
         "ecdh_curves": { "type": "array", "items": { "type": "string" }  },
         "secret_namespacing": { "type": "boolean" },
         "redirect_cleartext_from": { "type": "integer" },
-        "sni": { "type": "string" }
+        "sni": { "type": "string" },
+        "verify_certificate_spki": { "type": "array", "items": { "type": "string" } },
+        "verify_certificate_hash": { "type": "array", "items": { "type": "string" } },
+        "verify_subject_alt_name": { "type": "array", "items": { "type": "string" } }
```
Comment (kflynn) [inline python/schemas/v1/TLSContext.schema]: The `v1` schemas are frozen, so no need to change this -- just leave it as it was. 🙂

Comment (jwm) [reply python/schemas/v1/TLSContext.schema]: Done!

Comment (jwm) [reply python/schemas/v1/TLSContext.schema]: Also, wasn't sure if you were going to take a crack at the tests; the changes in this PR don't meaningfully test these changes.

**Comment by kflynn** [comment]
@jwm, sorry for the delay here! Can you take a crack at conflict resolution?

**Comment by jwm** [comment]
@kflynn no worries, done!

**Comment by kflynn** [comment]
Thanks! On testing, if you check out `python/tests/t_tls.py`, you'll find the `ClientCertificateAuthentication` test. I think that that'll show you everything you need to test it...

**Comment by jkornwitz-ptc** [comment]
Are there any updates for this or https://github.com/datawire/ambassador/issues/1543?

**Comment by imranismail** [comment]
Are there any updates to this?

**Comment by tsuna** [comment]
I'm going to join the chorus here and ask: can we get this merged please?

**Comment by jwm** [comment]
@kflynn this reminds me, after our conversation recently, I was going to refactor the `TLSContext` directives onto the Ambassador `Module`. Then, I think you were going to take a crack at the tests? I'll try to make time for the refactor next week.

**Review by kflynn — APPROVED** [review]
Taking this one -- thanks for sticking with it! 🙂

**Pull request merged**